File: app/src/test/java/com/dominikgaller/alpinebooking/booking/core/domain/GuideTourTest.java
package com.dominikgaller.alpinebooking.booking.core.domain;

import com.dominikgaller.alpinebooking.booking.core.domain.event.TourStarted;
import com.dominikgaller.alpinebooking.booking.core.domain.exception.InvalidGuideTourStateException;
import com.dominikgaller.alpinebooking.booking.core.domain.exception.TourStartTooEarlyException;
import com.dominikgaller.alpinebooking.shared.domain.event.DomainEvent;
import org.junit.jupiter.api.Test;

import java.time.Instant;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatExceptionOfType;

class GuideTourTest {

    private static final Instant SCHEDULED_START = Instant.parse("2026-06-15T09:00:00Z");
    private static final Instant AT_SCHEDULED_START = SCHEDULED_START;
    private static final Instant AFTER_SCHEDULED_START = Instant.parse("2026-06-15T09:05:00Z");
    private static final Instant BEFORE_SCHEDULED_START = Instant.parse("2026-06-15T08:55:00Z");

    private static final GuideTourId TOUR_ID = GuideTourId.generate();
    private static final TourId TOUR_REF = new TourId("TOUR-42");

    private GuideTour scheduledTour() {
        return GuideTour.schedule(TOUR_ID, TOUR_REF, SCHEDULED_START);
    }

    // ── Happy path ────────────────────────────────────────────────────────────

    @Test
    void start_transitionsToRunning_whenStartedAtEqualsScheduledStart() {
        final GuideTour tour = scheduledTour();

        tour.start(AT_SCHEDULED_START);

        assertThat(tour.status()).isEqualTo(GuideTourStatus.RUNNING);
    }

    @Test
    void start_transitionsToRunning_whenStartedAtIsAfterScheduledStart() {
        final GuideTour tour = scheduledTour();

        tour.start(AFTER_SCHEDULED_START);

        assertThat(tour.status()).isEqualTo(GuideTourStatus.RUNNING);
    }

    @Test
    void start_setsStartedAt() {
        final GuideTour tour = scheduledTour();

        tour.start(AFTER_SCHEDULED_START);

        assertThat(tour.startedAt()).isEqualTo(AFTER_SCHEDULED_START);
    }

    @Test
    void start_emitsTourStartedEvent() {
        final GuideTour tour = scheduledTour();

        tour.start(AFTER_SCHEDULED_START);

        final List<DomainEvent> events = tour.pullDomainEvents();
        assertThat(events).hasSize(1);
        assertThat(events.get(0)).isInstanceOf(TourStarted.class);
        final TourStarted event = (TourStarted) events.get(0);
        assertThat(event.guideTourId()).isEqualTo(TOUR_ID);
        assertThat(event.tourId()).isEqualTo(TOUR_REF);
        assertThat(event.startedAt()).isEqualTo(AFTER_SCHEDULED_START);
    }

    @Test
    void pullDomainEvents_returnsEmptyOnSecondCall() {
        final GuideTour tour = scheduledTour();
        tour.start(AFTER_SCHEDULED_START);
        tour.pullDomainEvents();

        assertThat(tour.pullDomainEvents()).isEmpty();
    }

    // ── TooEarly ──────────────────────────────────────────────────────────────

    @Test
    void start_throwsTourStartTooEarlyException_whenBeforeScheduledStart() {
        final GuideTour tour = scheduledTour();

        assertThatExceptionOfType(TourStartTooEarlyException.class)
                .isThrownBy(() -> tour.start(BEFORE_SCHEDULED_START));
    }

    @Test
    void start_tooEarly_doesNotChangeStatus() {
        final GuideTour tour = scheduledTour();

        try {
            tour.start(BEFORE_SCHEDULED_START);
        } catch (TourStartTooEarlyException ignored) {
        }

        assertThat(tour.status()).isEqualTo(GuideTourStatus.SCHEDULED);
    }

    // ── Invalid state ─────────────────────────────────────────────────────────

    @Test
    void start_throwsInvalidGuideTourStateException_whenAlreadyRunning() {
        final GuideTour tour = GuideTour.reconstitute(
                TOUR_ID, TOUR_REF, SCHEDULED_START, GuideTourStatus.RUNNING, AFTER_SCHEDULED_START);

        assertThatExceptionOfType(InvalidGuideTourStateException.class)
                .isThrownBy(() -> tour.start(AFTER_SCHEDULED_START));
    }

    @Test
    void start_throwsInvalidGuideTourStateException_whenFinished() {
        final GuideTour tour = GuideTour.reconstitute(
                TOUR_ID, TOUR_REF, SCHEDULED_START, GuideTourStatus.FINISHED, AFTER_SCHEDULED_START);

        assertThatExceptionOfType(InvalidGuideTourStateException.class)
                .isThrownBy(() -> tour.start(AFTER_SCHEDULED_START));
    }

    @Test
    void start_throwsInvalidGuideTourStateException_whenCancelled() {
        final GuideTour tour = GuideTour.reconstitute(
                TOUR_ID, TOUR_REF, SCHEDULED_START, GuideTourStatus.CANCELLED, null);

        assertThatExceptionOfType(InvalidGuideTourStateException.class)
                .isThrownBy(() -> tour.start(AFTER_SCHEDULED_START));
    }

    // ── schedule factory ──────────────────────────────────────────────────────

    @Test
    void schedule_setsInitialStatusToScheduled() {
        final GuideTour tour = scheduledTour();

        assertThat(tour.status()).isEqualTo(GuideTourStatus.SCHEDULED);
    }

    @Test
    void schedule_hasNoPendingEvents() {
        final GuideTour tour = scheduledTour();

        assertThat(tour.pullDomainEvents()).isEmpty();
    }
}

File: src/test/kotlin/com/example/contractmanagement/guide/core/domain/guidetour/GuideTourTest.kt
package com.example.contractmanagement.guide.core.domain.guidetour

import com.example.contractmanagement.guide.core.domain.guidetour.exception.InvalidGuideTourStateException
import com.example.contractmanagement.guide.core.domain.guidetour.exception.TourStartTooEarlyException
import com.example.contractmanagement.shared.domain.TourId
import com.example.contractmanagement.shared.domain.event.TourStarted
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.Test
import java.time.Instant
import java.util.UUID

/**
 * Domain tests for the [GuideTour] aggregate.
 *
 * SDD: see `documentation/domain/aggregate-guide-tour.spec.md`.
 */
class GuideTourTest {
    private val uuid: UUID = UUID.fromString("33333333-3333-3333-3333-333333333333")
    private val id = GuideTourId(uuid)
    private val tourId = TourId("TOUR-42")
    private val scheduledStart: Instant = Instant.parse("2026-06-01T09:00:00Z")
    private val onTime: Instant = Instant.parse("2026-06-01T09:05:00Z")

    private fun scheduled(): GuideTour = GuideTour.schedule(id, tourId, scheduledStart)

    private fun inState(
        status: GuideTourStatus,
        startedAt: Instant? = null,
    ): GuideTour =
        GuideTour.reconstitute(
            id = id,
            tourId = tourId,
            scheduledStart = scheduledStart,
            status = status,
            startedAt = startedAt,
        )

    @Test
    fun schedule_setsStatus_toScheduled() {
        assertThat(scheduled().status).isEqualTo(GuideTourStatus.SCHEDULED)
    }

    @Test
    fun schedule_recordsNoEvents() {
        assertThat(scheduled().pullDomainEvents()).isEmpty()
    }

    @Test
    fun schedule_leavesStartedAtAbsent() {
        assertThat(scheduled().startedAt).isNull()
    }

    @Test
    fun start_transitionsStatus_toRunning() {
        val tour = scheduled()

        tour.start(onTime)

        assertThat(tour.status).isEqualTo(GuideTourStatus.RUNNING)
    }

    @Test
    fun start_recordsTheActualStartTime() {
        val tour = scheduled()

        tour.start(onTime)

        assertThat(tour.startedAt).isEqualTo(onTime)
    }

    @Test
    fun start_publishesTourStarted_carryingTheIdentityAsAString() {
        val tour = scheduled()

        tour.start(onTime)

        assertThat(tour.pullDomainEvents())
            .containsExactly(TourStarted(uuid.toString(), tourId, onTime))
    }

    @Test
    fun start_succeeds_whenStartedExactlyAtTheScheduledTime() {
        val tour = scheduled()

        tour.start(scheduledStart)

        assertThat(tour.status).isEqualTo(GuideTourStatus.RUNNING)
    }

    @Test
    fun start_throwsTourStartTooEarlyException_whenBeforeTheScheduledTime() {
        val tour = scheduled()

        assertThatThrownBy { tour.start(scheduledStart.minusSeconds(1)) }
            .isInstanceOf(TourStartTooEarlyException::class.java)
    }

    @Test
    fun start_leavesStateUntouched_whenTooEarly() {
        val tour = scheduled()

        runCatching { tour.start(scheduledStart.minusSeconds(1)) }

        assertThat(tour.status).isEqualTo(GuideTourStatus.SCHEDULED)
        assertThat(tour.startedAt).isNull()
        assertThat(tour.pullDomainEvents()).isEmpty()
    }

    @Test
    fun start_throwsInvalidGuideTourStateException_whenAlreadyRunning() {
        val tour = inState(GuideTourStatus.RUNNING, startedAt = onTime)

        assertThatThrownBy { tour.start(onTime) }
            .isInstanceOf(InvalidGuideTourStateException::class.java)
    }

    @Test
    fun reconstitute_recordsNoEvents() {
        assertThat(inState(GuideTourStatus.RUNNING, onTime).pullDomainEvents()).isEmpty()
    }
}
